Introduce a runtimes caching and management layer

chandlerc opened the pull request:
This layer allows runtimes to be built on-demand but cached in a consistent and re-usable location on the system. It handles careful filesystem operations to ensure consistency even in the face of multiple versions and build configurations.

This addresses a number of TODOs from the initial runtimes building on-demand, and sets the stage to scale up to more runtimes.

This doesn't switch on-demand runtimes to be on by default, I wanted to wait and make that change as a separate step.

chandlerc committed:
Introduce a runtimes caching and management layer

This layer allows runtimes to be built on-demand but cached in
a consistent and re-usable location on the system. It handles careful
filesystem operations to ensure consistency even in the face of multiple
versions and build configurations.

This addresses a number of TODOs from the initial runtimes building
on-demand, and sets the stage to scale up to more runtimes.

This doesn't switch on-demand runtimes to be on by default, I wanted to
wait and make that change as a separate step.

chandlerc committed:
Improve building of generated sources for ClangD

We have grown more generation rules, so try to use a regex instead of
listing all of them. Also include the runfiles logic that isn't
generated, but needs to be built to be present in the tree.

Also, add a filtering step to remove extra compilations from the
database for the host or "exec" configuration in Bazel. These end up
being duplicate compilations of the same source file. The different
compilation flags would be fine, but they also use a different Bazel
tree for generated files, and one that we can't run a Bazel command to
populate with those generated files easily the way we can and do for
the main configuration. Filtering these out still leaves the main
configuration.

This will likely both make things like `clangd-tidy` even faster by
avoiding redundant work in multiple configurations, and it avoids errors
when those generated files are missing.

The downside of this change is that it means we will need multiple
clangd-tidy actions to build the compilation database in different
configurations in order to see clang-tidy warnings that only occur in
that configuration. This doesn't seem to be a major issue, and we can
pursue it in the future if needed.

danakj commented:
@geoffromer Looks like this was assigned to you first, I can leave it for you to keep reviewing?

chandlerc commented:
FWIW, while I patched in a hopeful clangd-tidy fix, it isn't yet sufficient. Would appreciate review on the rest as I can land it independently, it's just the clangd-tidy check will fail because of the issue there. I've found and fixed the actual warning it encountered so this shouldn't really be blocking anything.

geoffromer commented:
> @geoffromer Looks like this was assigned to you first, I can leave it for you to keep reviewing?

Sure, will do.

geoffromer commented on toolchain/driver/runtimes_cache.h:
```diff
@@ -0,0 +1,350 @@
+//
+// Beyond providing access, this class also supports orchestrating the _build_
+// of these runtimes, including synchronizing between different threads or
+// processes trying to build the same component of the runtimes.
```
I'm having a hard time understanding the structural relationships here. I'm guessing a "runtime" is the same thing as a "runtime library", but is there a one-to-one relationship between runtime libraries and runtime directories, and/or components, or do they nest (e.g. one component has multiple runtimes), or are they just separate?
Also, what's the relationship between runtimes in the same `Runtimes`, vs. runtimes in different `Runtimes` in the same `Runtimes::Cache`?

I think part of the problem is that sometimes "runtimes" is the plural of "runtime", but other times it's a singular noun that means "an object of type `Runtimes`", and it can be hard to tell which is which. Would it make sense to always use "Runtimes" (capital "R") rather than "runtimes" to refer to an instance of the type?

chandlerc replied:
Reworded this to try and help. Also see below.

geoffromer commented on toolchain/driver/runtimes_cache.h:
```diff
@@ -0,0 +1,350 @@
+
+// A managed cache of runtimes.
+//
+// This class manages and provides access to a cache of runtimes. The runtimes
```
Here too, I'm having trouble understanding the structural relationships. The comments describe this as a cache of runtimes (i.e. a cache where each element is a runtime), but the signature of `Lookup` strongly suggests that this is a cache of `Runtimes`s (i.e. a cache where each element is some sort of collection or aggregate of runtimes). Assuming the latter interpretation is correct, what's the significance of the two levels of nesting?

Part of the problem is that sometimes "runtimes" is the plural of "runtime", but other times it's a singular noun that means "an object of type `Runtimes`", and it can be hard to tell which is which. Would it be practical to use "Runtimes" (capital "R") rather than "runtimes" to refer to an instance of the type, and something like "`Runtimes`s" to refer to more than one of them?

chandlerc replied:
I think the key is that a single entry in the cache is a set of runtimes, not a singular runtime.

I've reworded this section to try and explain this, and introduce the informal term of "runtimes" for each entry. Let me know if this helps?

geoffromer replied:
> I think the key is that a single entry in the cache is a set of runtimes, not a singular runtime.

And yet it's repeatedly referred to as a singular runtime, e.g. "If a runtime for these features does not exist in the cache" on line 250, or "Commits the new runtime to the cache" on line 321. In fact, I can't find any uses of the singular noun "runtime" that _don't_ refer to a whole cache entry, and only a few incidental uses of "runtimes" as a plural noun that don't refer to multiple whole cache entries (although there are also a lot of uses of "runtimes" as a _singular_ noun, to refer to a `Runtimes` object).

So it seems like there's a more consistent terminology within reach:
- An entry in the cache is called a "runtime" (and/or a "runtime tree" or "runtime directory"), and the plural "runtimes" relates to the singular in the usual way.
- A sub-directory of a runtime is called a "runtime library" or "component".

To get there, I think we'd need to rename `Runtimes` to `Runtime`, change "runtimes" to "runtime" when it's referring to an instance of the class (which will fix "an unformed runtimes" and similar grammatical awkwardness), and change a couple of occurrences of "runtime" to "runtime library" in the second half of the file. If you like I can send you a PR against your branch to show you what I mean.

Would that be inconsistent with the conventional meaning of "a runtime"?

> I've reworded this section to try and explain this, and introduce the informal term of "runtimes" for each entry. Let me know if this helps?

Yeah, with the more explicit explanation in hand, I can mostly work out whether a given use of "runtimes" refers to one entry or multiple, and whether a given use of "runtime" refers to a whole entry or part of one. It would be nice not to have to do that, though (and it might be more confusing to someone who's not reading the file from top to bottom).

chandlerc replied:
> > I think the key is that a single entry in the cache is a set of runtimes, not a singular runtime.
> 
> And yet it's repeatedly referred to as a singular runtime, e.g. "If a runtime for these features does not exist in the cache" on line 250, or "Commits the new runtime to the cache" on line 321. In fact, I can't find any uses of the singular noun "runtime" that _don't_ refer to a whole cache entry, and only a few incidental uses of "runtimes" as a plural noun that don't refer to multiple whole cache entries (although there are also a lot of uses of "runtimes" as a _singular_ noun, to refer to a `Runtimes` object).

I'm happy to fix places where I've used a singular `runtime` as I think that is a confusing term to use -- these really are *sets* of runtimes.

And we do use it in a singular sense correctly in for example the comment on the `Runtimes` class where we talk about an individual "runtime library" within that "runtimes" tree.

> 
> So it seems like there's a more consistent terminology within reach:
> 
> * An entry in the cache is called a "runtime" (and/or a "runtime tree" or "runtime directory"), and the plural "runtimes" relates to the singular in the usual way.
> * A sub-directory of a runtime is called a "runtime library" or "component".
> 
> To get there, I think we'd need to rename `Runtimes` to `Runtime`, change "runtimes" to "runtime" when it's referring to an instance of the class (which will fix "an unformed runtimes" and similar grammatical awkwardness), and change a couple of occurrences of "runtime" to "runtime library" in the second half of the file. If you like I can send you a PR against your branch to show you what I mean.
> 
> Would that be inconsistent with the conventional meaning of "a runtime"?

Yes, IMO.

Unfortunately, a singular "runtime" as a noun has a somewhat different but somewhat overlapping meaning: I would typically assume that referred to something like the JVM or C-Python -- *the* runtime for a language.

Here, we're talking about runtime _libraries_ and runtime _data_. Together, all the parts of this that are linked into a C++ program would constitute the "runtime" of that program, but referring to the collection of *available* libraries and data as the "runtime" would be confusing to me.

While the change to being plural is small, for me it helps a lot to not assume what *specific* usage of "runtime" we might be talking about within the collection, and then the documentation for that specific component can clarify.

I'd prefer to talk about the "runtimes", a "runtimes component", and a "runtime library".

The `Runtimes` class should be the first, the Clang resource directory is an example of the second, and the builtins library within the resource directory an example of the third.

I'm using that terminology as I implement the API adjustments suggested.

geoffromer replied:
OK. I think if I were writing this I'd avoid using "runtimes" unqualified, and instead use whichever of "runtimes tree", "runtimes directory", "`Runtimes`", "cache entry", or "cache entries" is most appropriate to the context. That may be unnecessarily pedantic, so I'm not insisting on it, but I've left some optional suggested edits in places where I think a more specific term would be particularly helpful. Feel free to resolve the optional edits without applying them if you prefer.

I've also suggested edits in places that use singular "runtime", or plural "runtimes" that doesn't line up with my understanding of how you're using that term. If those places are right then I'm still misunderstanding something, so they're not marked optional.

chandlerc committed:
sync compdb with trunk

chandlerc committed:
improve comments

geoffromer commented on toolchain/driver/runtimes_cache.h:
```diff
@@ -0,0 +1,359 @@
+  // The base directory for the runtimes.
+  auto base_dir() const -> Filesystem::DirRef { return base_dir_; }
+
+  // Gets the path to an _existing_ Clang resource directory.
```
Is the Clang resource directory one of the runtime libraries managed by this object? Is the idea that when we add more runtime libraries, each one will have its own Get/Build method pair? Might it instead make sense to rename these e.g. `GetExistingLibrary` and have them take the subdirectory name as a parameter? If that works it would not only avoid boilerplate, it would also be much clearer about how the API relates to the concepts established in the header (like "runtime library").

chandlerc replied:
It's a bit awkward.

The resource directory is in some ways a *collection* of runtime resources, but it is one that is required to be in a single directory with a specific internal layout. The result is that we need to treat it as a single "component" here -- the whole thing needs to be built as one step.

Each component of the runtimes that is built separately will need its own Get/Build pair. I didn't want to try and design the API for multiple of them until I had examples of multiple, but I guess I can go ahead and refactor it that way now with a single entry.

I want to avoid necessarily calling these "library" though because they aren't always a library at all or aren't a singular library. The granularity stems from how independently they can be *built*, not from their contents or how they are consumed.

geoffromer commented on toolchain/driver/runtimes_cache.h:
```diff
@@ -0,0 +1,359 @@
+  static constexpr auto MaxLockedEntryAge = std::chrono::days(10);
+
+  // Runtimes are locked while in use to avoid them being removed concurrently,
+  // but this is an even more advisory lock than usual. We use a relatively
```
```suggestion
  // but the lock will be disregarded for entries older than `MaxLockedEntryAge`. We use a relatively
```

geoffromer commented on toolchain/driver/runtimes_cache.cpp:
```diff
@@ -0,0 +1,496 @@
+    // As a debugging aid, write our PID into the lock file when we
+    // successfully acquire it. Ignore errors here though.
+    auto write_result = lock_file.WriteFileFromString(std::to_string(getpid()));
+    static_cast<void>(write_result);
```
Optional:

```suggestion
    (void)write_result;
```

Both forms seem to occur in our codebase with about the same frequency, but I find this one easier to read. With that change, it might also make sense to drop `write_result` and discard the function return value directly.

chandlerc replied:
Sure, done.

geoffromer commented on toolchain/driver/runtimes_cache.cpp:
```diff
@@ -0,0 +1,496 @@
+  }
+
+  // Otherwise, we will need to build the runtimes and commit them into this
+  // directory once ready. Try and acquire an advisory lock to avoid redundant
+  // computation.
```
How does this avoid redundant computation, though, when this function builds the runtimes regardless of whether the lock is acquired?

chandlerc replied:
Doh, I just forgot to re-check for an existing path. Added.

geoffromer commented on toolchain/driver/runtimes_cache.cpp:
```diff
@@ -0,0 +1,496 @@
+  // Carbon are cached, persistent built library data, not something that only
+  // exists during the running of the Carbon tool like a socket.
+  //
+  // If we find a cache directory, we locate or create a Carbon specific
```
Grammar nit:
```suggestion
  // If we find a cache directory, we locate or create a Carbon-specific
```

geoffromer commented on toolchain/driver/runtimes_cache.cpp:
```diff
@@ -0,0 +1,496 @@
+    if (const char* xdg_cache_home = getenv("XDG_CACHE_HOME");
+        xdg_cache_home != nullptr) {
+      path_ = xdg_cache_home;
+      CARBON_VLOG("Using '$XDG_CACHE_HOME' cache: {0}", path_);
```
This log entry seems like it could be confusing in cases where `path_` is empty or relative, so it's not actually used.

chandlerc replied:
Fixed to be specific on when it is being used.

geoffromer commented on toolchain/driver/runtimes_cache.cpp:
```diff
@@ -0,0 +1,496 @@
+      }
+    }
+    if (!path_.empty()) {
+      // We have a candidate XDG-based cache path. Try to open that, and a
```
This doesn't seem quite right: we might instead have a `$HOME`-based path at this point, right?

chandlerc replied:
XDG actually specifies the fallback to `$HOME/.cache`. The difference is between an explicit `$XDG_CACHE_HOME` environment variable and the fallback based on `$HOME` -- both are specified by XDG.

geoffromer commented on toolchain/driver/runtimes_cache.cpp:
```diff
@@ -0,0 +1,496 @@
+        }
+      }
+    }
+    if (!path_.empty()) {
```
It looks like if the XDG-based path is relative, this branch will be entered `$HOME` isn't set, but won't be entered if `$HOME` is set but relative. It seems like at least one of those behaviors must be a bug (probably the first).

chandlerc replied:
I think all of this is fixed with the restructuring.

geoffromer commented on toolchain/driver/runtimes_cache.cpp:
```diff
@@ -0,0 +1,496 @@
+    // If we have a digest to use as the cache key, save it and we can try to
+    // use persistent caches.
+    cache_key_ = *std::move(read_digest_result);
+
```
I wonder if it would be clearer to factor out a couple of free functions here, in order to rely more on function calls and early returns rather than implicit fall-through and `path_` as a sentinel. For example:
- Factor out the final directory-opening logic as a free function `OpenPersistentCacheAt` that takes a path and returns a `StatusOr<Dir>`, with an explicit early return if the path doesn't exist.
- Factor out everything else from here to the end of the scope as a free function that returns a `StatusOr<Dir>`, implemented in terms of separate `return OpenPersistentCacheAt(...);` statements in the XDG and `$HOME` branches.
- Here, initialize `dir_`, `dir_owner_`, and `path_` using the `Dir` returned by the previous function (or propagate the failure if there was one), possibly sharing code with the `MakeTmpDir` case that follows.

That seems like it would help avoid the issues I've flagged below, as well as making the comment about fall-through unnecessary (and of course breaking up the long function body).

chandlerc replied:
I've gone with a slightly different factoring that seemed a bit easier to manage.

One of the annoying things is that this code tries to propagate some "fatal" errors, and falls back to using a temporary cache for "non-fatal" errors. =/ That said, I think I found a reasonable way to break it apart some.

geoffromer commented on toolchain/driver/runtimes_cache.cpp:
```diff
@@ -0,0 +1,496 @@
+        // create a subdirectory if one doesn't yet exist.
+        path_ /= std::string_view(CachePath);
+        CARBON_ASSIGN_OR_RETURN(
+            dir_owner_,
```
Doesn't this mean that the cache will be removed from the filesystem before the end of execution? If so, what's "persistent" about it?

chandlerc replied:
The owner can either be an open directory, or a removing directory. Here, we're just opening the directory and so there won't be any removal.

The ownership is of the open directory handle.

geoffromer commented on toolchain/driver/runtimes_cache.h:
```diff
@@ -0,0 +1,359 @@
+  static constexpr auto RuntimesLockDeadline = std::chrono::milliseconds(100);
+  static constexpr auto RuntimesLockPollInterval = std::chrono::milliseconds(1);
+
+  explicit Cache(llvm::raw_ostream* vlog_stream) : vlog_stream_(vlog_stream) {}
+
+  auto InitSystemCache(const InstallPaths& install) -> ErrorOr<Success>;
+  auto InitCachePath(const InstallPaths& install,
+                     std::filesystem::path cache_path) -> ErrorOr<Success>;
```
It would help to have comments on these methods, either here or on their definitions.

chandlerc replied:
Added comments. Note that these are just factoring the logic from the factory functions into methods so that we can use `CARBON_VLOG`. The relevant comment is really the relevant `Make*` comment, and I've documented that.

As I ended up splitting out functions though, I've added more substantive comments to those.

geoffromer commented on toolchain/driver/runtimes_cache.cpp:
```diff
@@ -0,0 +1,496 @@
+}
+
+auto Runtimes::Cache::Lookup(const Features& features) -> ErrorOr<Runtimes> {
+  // Compute the hash of the features, we'll use this to build the subdirectory
```
Grammar nit:
```suggestion
  // Compute the hash of the features. We'll use this to build the subdirectory
```

geoffromer commented on toolchain/driver/runtimes_cache.cpp:
```diff
@@ -0,0 +1,496 @@
+    // Try to get an `mtime` from the lock file in the directory. We don't use
+    // the directory itself to avoid any oddities with `mtime` on directories.
+    // Note that we ignore errors here as if we can't read the stamp file we
+    // will pick an arbitrary old time stamp.
```
This comment seems like it's just saying in English what the `stat_result.ok() ? ...` below says in pretty self-explanatory C++, but the "note that we ignore..." phrasing makes me worry you're trying to point out something subtle that I missed. If the former, this sentence should probably be removed; if the latter, it should probably be rephrased.

chandlerc replied:
There are two subtleties here, I've tried to rework the comment to make this more clear.

geoffromer commented on toolchain/driver/runtimes_cache.cpp:
```diff
@@ -0,0 +1,496 @@
+    Filesystem::ReadFile lock_file;
+    Filesystem::FileLock flock;
+    if (age <= MaxLockedEntryAge) {
+      CARBON_VLOG("Attempting to lock cache entry '{0}'", name);
```
The control flow here is hard for me to follow, because of the way it mixes fall-through with `continue` short-circuiting and a couple different state mutations, across several layers of nesting. It seems like it would help to factor out a separate function for acquiring the lock, from here to the end of the scope, although the need to treat a missing lock file as a successful removal makes that a little messy.

chandlerc replied:
Yeah, sorry about that.

Factoring this entirely out into a separate function is a little awkward, but it seems easy to have another helper lambda that provides the nicer early-exit logic. Does that work for you?

geoffromer commented on toolchain/driver/runtimes_cache.cpp:
```diff
@@ -0,0 +1,496 @@
+        CARBON_VLOG("Error opening lock file for cache entry '{0}': {1}", name,
+                    lock_file_open_result.error());
+        continue;
+      } else {
```
I think it would be clearer to drop the `else` here, i.e. structuring the lock acquisition as 
```
<begin happy path>
if (<had error>) {
  <handle error>;
  continue;
}
<rest of happy path>
```
instead of
```
<begin happy path>
if (<had error>) {
  <handle error>;
  continue;
} else {
  <rest of happy path>
}
```

That helps distinguish the happy path from the error cases by keeping its pieces at the same nesting level.

chandlerc replied:
I think the factored code looks like this.

geoffromer commented on toolchain/driver/runtimes_cache.cpp:
```diff
@@ -0,0 +1,496 @@
+                              RuntimesLockDeadline, RuntimesLockPollInterval);
+        if (lock_result.ok()) {
+          flock = *std::move(lock_result);
+        } else {
```
Same basic idea here: consider instead
```
if (!lock_result.ok()) {
  <handle error>
  continue;
}
flock = *std::move(lock_result);
```

chandlerc replied:
And here.

geoffromer commented on toolchain/driver/runtimes_cache.cpp:
```diff
@@ -0,0 +1,496 @@
+      }
+
+      // We should always have acquired the lock at this point.
+      CARBON_CHECK(flock.is_locked());
```
I suspect that if you implement my previous two suggestions, this will be obvious, because it will come immediately after `flock = *std::move(lock_result);`, with no intervening control flow.

chandlerc replied:
Yep.

geoffromer commented on toolchain/driver/runtimes_cache.cpp:
```diff
@@ -0,0 +1,496 @@
+  // release that state.
+  if (rename_result.ok()) {
+    std::move(dir_).Release();
+  } else if (!rename_result.error().not_empty()) {
```
Consider structuring this like
```
} else if (<error is not-empty>) {
  <handle not-empty error>
} else {
  <handle any other error>
}
```
instead of
```
} else if (<error is not not-empty>) {
  <handle any other error>
} else {
  <handle not-empty error>
}
```

That way you avoid the double negation, and the blocks are more straightforwardly related to their conditions (the block that handles not-empty comes after the condition that mentions not-empty, and the block that handles everything else comes after `else`). It's also a nice bonus that the two potentially-falling-through cases are grouped together, and the two potentially-returning cases are grouped together.

chandlerc replied:
Sure.

geoffromer commented on toolchain/driver/runtimes_cache.cpp:
```diff
@@ -0,0 +1,496 @@
+    // them.
+    CARBON_VLOG("PID {0} found racily built runtimes in cache path: {1}",
+                getpid(), dest_path);
+    CARBON_RETURN_IF_ERROR(std::move(dir_).Remove());
```
Consider using ordinary control flow here instead, because I'm finding that `CARBON_RETURN_IF_ERROR` is making the code harder for me to follow. I'm used to it being a way to avoid disrupting the happy-path logic with error handling branches, but here we're already off the happy path in branching-on-errors mode, and having different branches use different idioms for returning errors makes it harder to take in the three-way branch as a whole.

chandlerc replied:
I don't really follow this...

This path is a perfectly happy path, we just lost the race to commit the runtimes here?

It's only the other case that is truly off the happy-path.

I can switch, but the code will literally be the body of this macro, which seems to add quite a bit of verbosity with no added information.

geoffromer replied:
> This path is a perfectly happy path, we just lost the race to commit the runtimes here?
>
> It's only the other case that is truly off the happy-path.

It's a bit of a gray area -- there is an error as far as the filesystem API is concerned, it's just not an error at the application level. Whether you consider this a happy path or a funny sort of error recovery, the issue for me was that it _looks like_ an error-handling path because the branch condition is `rename_result.error().not_empty()`. I think that appearance, combined with the absence of error control flow macros in the rest of the function (even in the true-error case), cued me to read the code in explicit control-flow "mode", and so for example this line was easy to overlook when skimming for things like early returns.

Maybe the fact that the true-error case is now below this line will help? I dunno, I've thought about it too much to read the new version with fresh eyes.

> I can switch, but the code will literally be the body of this macro, which seems to add quite a bit of verbosity with no added information.

Yeah, it's a tradeoff for sure. On net I think it would be more readable for me personally, but others' mileage may vary. It's your call.

chandlerc replied:
I'm really benefiting from the macros here, so I'd prefer to keep them I think. I understand its a close call and varies, but the verbosity was a pretty major challenge for me.

geoffromer commented on toolchain/driver/runtimes_cache.cpp:
```diff
@@ -0,0 +1,496 @@
+    std::filesystem::path dest_path = runtimes_->base_path() / dest_;
+    CARBON_VLOG("PID {0} releasing lock on cache path: {1}", getpid(),
+                dest_path);
+    // FIXME: better write API
```
```suggestion
    // TODO: better write API
```

chandlerc replied:
No, this is stale -- the better write API is now used and I just needed to clean up the FIXME.

geoffromer commented on toolchain/driver/runtimes_cache.cpp:
```diff
@@ -0,0 +1,496 @@
+                dest_path);
+    // FIXME: better write API
+    auto write_result = lock_file_.WriteFileFromString("");
+    static_cast<void>(write_result);
```
```suggestion
    (void)write_result;
```

chandlerc replied:
Done (but directly on the write).

geoffromer commented on toolchain/driver/runtimes_cache.cpp:
```diff
@@ -0,0 +1,496 @@
+    static_cast<void>(write_result);
+    flock_ = {};
+    auto close_result = std::move(lock_file_).Close();
+    static_cast<void>(close_result);
```
```suggestion
    (void)close_result;
```

chandlerc replied:
Done (but directly on the write)

geoffromer commented on toolchain/driver/runtimes_cache_test.cpp:
```diff
@@ -0,0 +1,717 @@
+  auto home_cache_dir = *test_home.CreateDirectories(".cache");
+  std::filesystem::path home_cache_path = home_path / ".cache";
+
+  // Save the environment variables we'll override for testing and restore them
+  // afterward to avoid test-to-test oddities.
+  constexpr const char* XdgCacheEnv = "XDG_CACHE_HOME";
+  constexpr const char* HomeEnv = "HOME";
+  const char* orig_xdg_cache = getenv(XdgCacheEnv);
+  const char* orig_home = getenv(HomeEnv);
+  auto restore_env = llvm::make_scope_exit([&] {
+    for (const auto [env, orig] : {std::pair{XdgCacheEnv, orig_xdg_cache},
+                                   std::pair{HomeEnv, orig_home}}) {
+      if (orig) {
+        setenv(env, orig, /*overwrite*/ true);
+      } else {
+        unsetenv(env);
+      }
+    }
+  });
```
Maybe put this at the top of the test, so it doesn't interrupt the test logic?

chandlerc replied:
Everything above this is non-test logic. I've moved the comments around to try and make that clear.

geoffromer commented on toolchain/driver/runtimes_cache_test.cpp:
```diff
@@ -0,0 +1,717 @@
+  EXPECT_THAT(result->path(), Not(StartsWith(home_cache_path)));
+  EXPECT_THAT(result->path(), Not(StartsWith(xdg_path)));
+
+  // Now the main XDG cache logic should work.
```
Why "now"? Why wouldn't it have worked before this?

chandlerc replied:
Clarified.

geoffromer commented on toolchain/driver/runtimes_cache_test.cpp:
```diff
@@ -0,0 +1,717 @@
+  // Remove the XDG cache directory, but leave the environment set. We want to
+  // be robust against this, but it isn't important *how* the fallback occurs,
+  // it could go to `$HOME/.cache`, or to a temporary directory.
+  ASSERT_THAT(tmp_dir_.Rmtree("xdg_cache_home"), IsSuccess(_));
```
Optional: I find tests easier to read if they only use googletest assertions when a failure would indicate a bug in the code under test, and use `CHECK` for everything else. That makes it easier to follow the intent of the test.

chandlerc replied:
Sure. I've fixed the places I saw in this test, but feel free to flag any I've missed.

geoffromer commented on toolchain/driver/runtimes_cache_test.cpp:
```diff
@@ -0,0 +1,717 @@
+  EXPECT_TRUE(*test_home.Access(".cache"));
+
+  // Now check a bunch of different failure modes for the home directory
+  // fallback. These should all end up creating temporary directories which
+  // we'll test functionally at the end.
```
If I'm right that `MakeSystem` will use a relative XDG path if `$HOME` is not set (see comment in the .cpp file), consider adding test coverage for that case. I'm not sure it's worth doing the kind of exhaustive testing that would have caught this beforehand, but it's probably worth at least covering the particular corner cases where we know we've had bugs.

chandlerc replied:
Sorry I missed this side of the comment, but I think if there was a way to use a relative path here it has been removed. I've even simplified the code a bit as `path.is_absolute()` implies `!path.empty()`. There should be no way to get to a relative case here.

geoffromer commented on toolchain/driver/runtimes_cache_test.cpp:
```diff
@@ -0,0 +1,717 @@
+
+  // Finally, double check that these temporary caches still produce a writable
+  // directory.
+  result = Runtimes::Cache::MakeSystem(install_);
```
I'm finding it hard to tell what state the system is in at each assertion (e.g. to figure out whether this test covers the case where I suspect a bug), because there are so many incremental state changes over the course of the test. Have you considered splitting this test into a bunch of smaller tests that each test a single behavior?

chandlerc replied:
Yes, I thought about doing that.

My hesitation is that it requires a lot of code to set up each of the initial states, and it didn't seem worth the added complexity to factor the setup code into a carefully re-usable form that allows incrementally building each step.

If it's really too hard to read in the current form, I can switch it, but it will make the test code longer and take more time, so I didn't want to do it without cause.

geoffromer commented on toolchain/driver/runtimes_cache_test.cpp:
```diff
@@ -0,0 +1,717 @@
+  llvm::SmallVector<std::filesystem::path> built_runtimes_paths;
+
+  for (const std::string& target : targets) {
+    auto lookup_result = cache_.Lookup({.target = target});
```
```suggestion
    SCOPED_TRACE(target);
    auto lookup_result = cache_.Lookup({.target = target});
```

(So that failure messages tell you which iteration the failure happened on)

geoffromer commented on toolchain/driver/runtimes_cache_test.cpp:
```diff
@@ -0,0 +1,717 @@
+
+  for (const auto& [target, built_runtimes_path] :
+       llvm::zip(targets, built_runtimes_paths)) {
+    auto lookup_result = cache_.Lookup({.target = target});
```
```suggestion
    SCOPED_TRACE(target);
    auto lookup_result = cache_.Lookup({.target = target});
```

geoffromer commented on toolchain/driver/runtimes_cache_test.cpp:
```diff
@@ -0,0 +1,717 @@
+      // result directly. Save it.
+      build2_path = std::get<std::filesystem::path>(*std::move(build_result2));
+    } else {
+      // In rare cases, the initial build will fail to acquire the file lock.
```
Does this mean that this branch will only be exercised rarely and unpredictably, and so failures in these assertions will be hard to reproduce? If so, I'm not sure they're adding much value. Do they provide any coverage that we don't also get from `ConcurrentBuildsWithFailedLocking`?

chandlerc replied:
Yes and IMO yes...

Note that while it's a bit hard to reproduce, I was able to find and fix some subtle bugs with --runs_per_test=N and Bazel on a single machine pretty easily.

In some ways, these unit tests, especially combined with `--runs_per_test=N` with large `N` end up forming nice stress tests that catch very odd sequences of operations.

geoffromer replied:
The comment on `ConcurrentBuildsWithFailedLocking` seems to say that it's just like `ConcurrentBuilds` except that it's designed to force this branch to occur more reliably. Am I misunderstanding something? If not, how do these assertions provide coverage that we don't get from the corresponding assertions in `ConcurrentBuildsWithFailedLocking`?

chandlerc replied:
Sorry, I think I understand better now your comment.

I see two main differences here. First and foremost is that when we induce the locking to fail below, we remove any coverage for actually having a thread block on the file lock. If that caused a crash in the system call, etc, that test will never catch it while this one (usually)  will.

Second, if the error isn't in the rare edge here but in the more normal case, this test will be much easier to debug than the one below because it should almost always occur rather than being a bit of a coin-toss.

I'd even be fine not really covering this edge in this test, but in practice even when *trying* to use the file lock, it sometimes fails and so we need the test to not misbehave.

geoffromer commented on toolchain/driver/runtimes_cache_test.cpp:
```diff
@@ -0,0 +1,717 @@
+  ASSERT_THAT(commit_result, IsSuccess(_));
+  std::filesystem::path build1_path = *std::move(commit_result);
+
+  // Even though there may be is another thread running, we should now get
```
"may be is" reads like an incomplete edit -- should it be "may be", or "is"?

chandlerc replied:
Fixed.

geoffromer commented on toolchain/driver/runtimes_cache_test.cpp:
```diff
@@ -0,0 +1,717 @@
+    ASSERT_THAT(commit_result, IsSuccess(_));
+    build2_path = *std::move(commit_result);
+
+    // Even though there may be another thread running, and even holding a lock
+    // file, we should now get non-blocking access directly to the built
+    // runtime.
```
It might be worth mentioning here that `ConcurrentBuildsLockTimeout` tests this scenario more rigorously (assuming that's true).

chandlerc replied:
I'm not sure which case is more rigorously tested here?

This is making sure that the current thread doesn't end up re-building here, or deadlock with both threads waiting on a lock file.

geoffromer replied:
> I'm not sure which case is more rigorously tested here?

I meant the case where there's another thread holding a lock file.

> This is making sure that the current thread doesn't end up re-building here, or deadlock with both threads waiting on a lock file.

But doesn't `EXPECT_THAT(runtimes2.Build(Runtimes::ClangResourceDir),...` in `ConcurrentBuildsLockTimeout` verify the same thing, while more reliably ensuring that the lock is held at this point?

chandlerc replied:
Ah, yes. I see what you're saying. Added a comment.

geoffromer commented on toolchain/install/make_installation_digest.cpp:
```diff
@@ -0,0 +1,132 @@
+
+ private:
+  llvm::raw_ostream* vlog_stream_ = nullptr;
+  Map<uint64_t, std::array<uint8_t, 32>> file_digests;
```
```suggestion
  Map<uint64_t, std::array<uint8_t, 32>> file_digests_;
```

chandlerc replied:
Done.

geoffromer commented on toolchain/install/make_installation_digest.cpp:
```diff
@@ -0,0 +1,132 @@
+namespace Carbon {
+namespace {
+
+// A class implementing our digest program.
```
Could there be a comment, here or on `main` or at the top of the file, that explains what this program does?

chandlerc replied:
Done.

geoffromer commented on toolchain/driver/build_runtimes_subcommand.cpp:
```diff
@@ -56,39 +55,44 @@ auto BuildRuntimesSubcommand::Run(DriverEnv& driver_env) -> DriverResult {
+  return {.success = true};
+}
+
+auto BuildRuntimesSubcommand::RunInternal(DriverEnv& driver_env)
```
Might it make sense to fold the `TestAndDiagnoseIfFuzzingExternalLibraries` call into this function? It seems like a clearer separation of concerns if `Run` itself does nothing but delegate to `RunInternal` and then log the result.

chandlerc replied:
The separation I was going for was instead between code that propagates errors in the return, versus code that uses the diagnostic infrastructure. The `RunInternal` code is the code that wants to propagate, and can use the macros that enable that.

The test and diagnose call does its own diagnostics and so needs to *not* propagate errors that way, which makes it hard to factor into `RunInternal`?

geoffromer commented on toolchain/driver/driver.cpp:
```diff
@@ -74,6 +78,34 @@ auto Options::Build(CommandLine::CommandBuilder& b) -> void {
+By default, the runtimes cache is located in the `carbon_runtimes` subdirectory
+of `$XDG_CACHE_HOME` (or `$HOME/.cache` if not set). If unable to use either, it
+will be placed in a temporary directory that is removed when the command
+completes. This flag overrides that logic with a specific path.
```
```suggestion
completes. This flag overrides that logic with a specific path. It has no effect
if --prebuilt-runtimes is set.
```

This wasn't completely obvious to me, so it seems worth spelling out.

geoffromer commented on toolchain/install/install_filegroups.bzl:
```diff
@@ -56,19 +58,24 @@ def install_target(name, target, executable = False, is_driver = False):
       is_driver: False if it should be included in the `no_driver_name`
         filegroup.
+      is_digest: False if it should be included in the `no_digest_name`
+        filegorup.
```
Nit:
```suggestion
        filegroup.
```

chandlerc committed:
Apply suggestions from code review

Co-authored-by: Geoff Romer <gromer@google.com>

chandlerc committed:
wip

chandlerc committed:
review-feedback

chandlerc reviewed (commented):
I think I got most of the feedback, thanks for the review. PTAL!

geoffromer commented on toolchain/driver/runtimes_cache.h:
```diff
@@ -268,18 +283,58 @@ class Runtimes::Cache {
 
+  // Tries to find a viable cache root.
+  //
+  // This must be an existing directory,
```
Nit: awkward line break.

chandlerc replied:
Fixed.

geoffromer commented on toolchain/driver/runtimes_cache.h:
```diff
@@ -0,0 +1,414 @@
+  friend class RuntimesTestPeer;
+
+  // The deadline for acquiring a lock to build a new component of the runtimes.
+  // This needs to be quite large as this is how long racing processes or
+  // threads will wait to allow some other process to complete building the
+  // runtimes. The result is that this should be significantly longer than the
+  // expected slowest-to-build runtimes.
+  //
+  // Note, nothing goes _wrong_ if this deadline is exceeded, but multiple
+  // copies of runtimes may end up being built and all but one thrown away.
+  static constexpr Filesystem::Duration BuildLockDeadline =
+      std::chrono::seconds(200);
+  // The interval at which to poll for a build lock. This needs to be small
+  // enough that we don't waste an excessive amount of time if a build of the
+  // runtimes completes *just* after a poll. Typically, that means we want this
+  // to be significant lower than the expected time it would take to build the
+  // runtimes. Note that we don't poll if the runtimes have been completely
+  // built prior to the query coming in, so this doesn't form the _minimum_ time
+  // to find a component of the runtimes tree.
```
Is this right?

```suggestion
  // This needs to be quite large as this is how long racing processes or
  // threads will wait to allow some other process to complete building the
  // component. The result is that this should be significantly longer than the
  // expected slowest-to-build component.
  //
  // Note, nothing goes _wrong_ if this deadline is exceeded, but multiple
  // copies of the component may end up being built and all but one thrown away.
  static constexpr Filesystem::Duration BuildLockDeadline =
      std::chrono::seconds(200);
  // The interval at which to poll for a build lock. This needs to be small
  // enough that we don't waste an excessive amount of time if a build of the
  // component completes *just* after a poll. Typically, that means we want this
  // to be significant lower than the expected time it would take to build the
  // component. Note that we don't poll if the component has been completely
  // built prior to the query coming in, so this doesn't form the _minimum_ time
  // to find a component of the runtimes tree.
```

geoffromer commented on toolchain/driver/runtimes_cache.h:
```diff
@@ -0,0 +1,414 @@
+  static constexpr Filesystem::Duration BuildLockPollInterval =
+      std::chrono::milliseconds(200);
+
+  // The path to the clang resource directory within the runtimes.
```
Optional:

```suggestion
  // The path to the clang resource directory within the runtimes tree.
```

geoffromer commented on toolchain/driver/runtimes_cache.h:
```diff
@@ -0,0 +1,414 @@
+  }
+
+  // A format string used to form the lock file for a given directory in the
+  // runtimes. This needs to be C-string, so directly expose the character
```
Optional:

```suggestion
  // runtimes tree. This needs to be C-string, so directly expose the character
```

geoffromer commented on toolchain/driver/runtimes_cache.h:
```diff
@@ -0,0 +1,414 @@
+// This class manages and provides access to a cache of runtimes. Each entry in
+// the cache is a collection of runtimes for a specific set of `Feature`s in a
+// directory that can be modeled with an object of the `Runtimes` type. We call
+// these entries "runtimes" typically. The cache keys and looks up these
```
```suggestion
// This class manages and provides access to a cache of runtimes. Each entry in
// the cache is a runtimes directory for a specific set of `Feature`s, represented by an
// object of the `Runtimes` type. The word "runtimes" can refer to a single entry or
// to multiple entries, depending on the context. The cache keys and looks up these
```

Two separate changes here, both optional:
- Using the previously-defined term "runtimes directory" rather than re-describing the thing it refers to. This is both for brevity and because avoiding that term gives the impression that it's somehow not applicable.
- Expanding the terminology clarification to be more explicit about the two meanings of "runtimes".

chandlerc replied:
The first part of this makes sense to me, but I think the second part isn't quite right...

The word "runtimes" should always refer to a single entry in the cache. The *contents* of that entry might be a single component or multiple components, but I think we always want to think of an entry as a collection of (potentially) multiple runtime components, which is why it is always plural. I'll make a change to somewhat emphasize this?

geoffromer replied:
OK, that change is helpful. I've flagged a few places where "runtimes" seems to refer to multiple cache entries, now that I know that's not intentional.

geoffromer commented on toolchain/driver/runtimes_cache.h:
```diff
@@ -0,0 +1,414 @@
+// used.
+class Runtimes::Cache {
+ public:
+  // The features of a cached runtime.
```
```suggestion
  // The features of a cached runtimes directory.
```

geoffromer commented on toolchain/driver/runtimes_cache.h:
```diff
@@ -0,0 +1,414 @@
+  // The path to the cache directory.
+  auto path() const -> const std::filesystem::path& { return path_; }
+
+  // Looks up a runtime in the cache.
```
```suggestion
  // Looks up a runtimes directory in the cache.
```

geoffromer commented on toolchain/driver/runtimes_cache.h:
```diff
@@ -0,0 +1,414 @@
+
+  // Looks up a runtime in the cache.
+  //
+  // This will return a `Runtimes` object for the given features. If a runtime
```
```suggestion
  // This will return a `Runtimes` object for the given features. If an entry
```

geoffromer commented on toolchain/driver/runtimes_cache.h:
```diff
@@ -0,0 +1,414 @@
+  // currently using that entry.
+  static constexpr auto MaxLockedEntryAge = std::chrono::days(10);
+
+  // Runtimes are locked while in use to avoid them being removed concurrently,
```
Optional:

```suggestion
  // Entries are locked while in use to avoid them being removed concurrently,
```

geoffromer commented on toolchain/driver/runtimes_cache_test.cpp:
```diff
@@ -0,0 +1,717 @@
+
+  // Commit the first built runtime.
+  auto commit_result = std::move(builder1).Commit();
+  ASSERT_THAT(commit_result, IsSuccess(_));
```
I think the thread underlying `build2_thread` will get leaked if this assertion fails. That's probably not a disaster, but did you consider using `std::jthread` instead?

chandlerc replied:
Doh, good call.

geoffromer reviewed (commented):
It looks like there are still a couple of unaddressed comments from the last round.

chandlerc committed:
Update toolchain/driver/runtimes_cache.h

Co-authored-by: Geoff Romer <gromer@google.com>

chandlerc committed:
Update toolchain/driver/runtimes_cache.h

Co-authored-by: Geoff Romer <gromer@google.com>

chandlerc committed:
Apply suggestions from code review

Co-authored-by: Geoff Romer <gromer@google.com>

chandlerc committed:
review

chandlerc committed:
more review

chandlerc reviewed (commented):
Hopefully converging a bit here, thanks!

geoffromer commented on toolchain/driver/runtimes_cache_test.cpp:
```diff
@@ -0,0 +1,707 @@
+      // In rare cases, the initial build will fail to acquire the file lock.
+      // The entire build process is designed specifically to be resilient to
+      // that so we should still succeed, but now we need to handle building in
+      // this thread as well.
```
Optional: I'm inclined to be slightly more explicit about the fact that this is an exception to the usual rule that flaky test failures are false positives.

```suggestion
      // this thread as well. Note that a true failure here may only
      // show up intermittently.
```

chandlerc replied:
I like that, thanks.

geoffromer commented on toolchain/driver/runtimes_cache.h:
```diff
@@ -0,0 +1,420 @@
+// old runtimes before creating the new ones. The eviction strategy is to remove
+// any runtimes more than a year old, as well as the least-recently used
+// runtimes until there will only be a maximum of 50 runtimes cached. The goal
+// is to allow multiple versions and build features to stay resident in the
```
Most if not all occurrences of "runtimes" in this passage appear to refer to multiple cache entries (i.e. they're synonyms for "entries", not "entry").

chandlerc replied:
Worse, I think this is a bit of a mix. I've fixed it to hopefully be more consistent.

geoffromer commented on toolchain/driver/runtimes_cache.h:
```diff
@@ -0,0 +1,420 @@
+//
+// These system-wide paths are only used if the installation contains a digest
+// file that can be used to ensure different builds and installs of Carbon don't
+// incorrectly share cached runtimes. When missing, a temporary directory is
```
```suggestion
// incorrectly share a cached runtimes. When missing, a temporary directory is
```

geoffromer commented on toolchain/driver/runtimes_cache.h:
```diff
@@ -0,0 +1,420 @@
+  auto ComputeEntryAges(llvm::SmallVector<std::filesystem::path> entry_paths)
+      -> llvm::SmallVector<Entry>;
+
+  // Prunes stale runtimes sufficiently to insert the provided new entry path
```
Here again, "runtimes" seems synonymous with "entries", not "an entry".

chandlerc replied:
Yeah, happy to say "entries" here. That seems more clear anyways as this is all about the caching, and not about the contents.

geoffromer reviewed (approved):
I'm happy for you to resolve the remaining comments without more feedback from me, unless you want it.

chandlerc committed:
Apply suggestions from code review

Co-authored-by: Geoff Romer <gromer@google.com>

chandlerc committed:
further comment improvements

chandlerc committed:
workaround jthread

chandlerc reviewed (commented):
Thanks so much for the review of a huge PR! Last comment fixes all made sense so should be good to go. Once CI is green I'll merge.

chandlerc committed:
one more comment

chandlerc committed:
Merge branch 'trunk' into runtimes-cache

chandlerc committed:
Merge branch 'trunk' into runtimes-cache

chandlerc committed:
Another attempt to fix ClangD-tidy

This restores the original approach in #6046 as it appears
`--notool_deps` isn't sufficient in some situations. It still isn't
clear to me why it seemed to work initially, but I can easily reproduce
the issue now even with that flag.

I've tried to address the feedback in the original PR on the Python
code.

chandlerc committed:
Merge branch 'trunk' into runtimes-cache
